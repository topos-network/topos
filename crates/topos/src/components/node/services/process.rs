use crate::edge::CommandConfig;
use std::collections::HashMap;
use std::path::PathBuf;
use std::process::ExitStatus;
use thiserror::Error;
use tokio::{spawn, sync::mpsc, task::JoinHandle};
use tokio_util::sync::CancellationToken;
use topos_config::sequencer::SequencerConfig;
use topos_config::tce::{AuthKey, StorageConfiguration, TceConfig};
use topos_p2p::Multiaddr;
use topos_sequencer::SequencerConfiguration;
use topos_tce_transport::ReliableBroadcastParams;
use topos_wallet::SecretManager;
use tracing::{debug, error, info, warn};

use topos_config::genesis::Genesis;

#[derive(Error, Debug)]
pub enum Errors {
    #[error("TCE error")]
    TceFailure,
    #[error("Sequencer error")]
    SequencerFailure,
    #[error("Edge error: {0}")]
    EdgeTerminated(#[from] std::io::Error),
}

pub fn generate_edge_config(
    edge_path: PathBuf,
    config_path: PathBuf,
) -> JoinHandle<Result<ExitStatus, Errors>> {
    // Create the Polygon Edge config
    info!("Generating the configuration at {config_path:?}");
    info!("Polygon-edge binary located at: {edge_path:?}");
    spawn(async move {
        CommandConfig::new(edge_path)
            .init(&config_path)
            .spawn()
            .await
            .map_err(|e| {
                error!("Failed to generate the edge configuration: {e:?}");
                Errors::EdgeTerminated(e)
            })
    })
}

pub(crate) fn spawn_sequencer_process(
    config: SequencerConfig,
    keys: &SecretManager,
    shutdown: (CancellationToken, mpsc::Sender<()>),
) -> JoinHandle<Result<ExitStatus, Errors>> {
    let config = SequencerConfiguration {
        subnet_id: config.subnet_id,
        public_key: keys.validator_pubkey(),
        subnet_jsonrpc_http: config.subnet_jsonrpc_http,
        subnet_jsonrpc_ws: config.subnet_jsonrpc_ws,
        subnet_contract_address: config.subnet_contract_address,
        tce_grpc_endpoint: config.tce_grpc_endpoint,
        signing_key: keys.validator.clone().unwrap(),
        verifier: 0,
        start_block: config.start_block,
    };

    debug!("Sequencer args: {config:?}");
    spawn(async move {
        topos_sequencer::run(config, shutdown).await.map_err(|e| {
            error!("Sequencer failure: {e:?}");
            Errors::SequencerFailure
        })
    })
}

pub(crate) fn spawn_tce_process(
    mut config: TceConfig,
    keys: SecretManager,
    genesis: Genesis,
    shutdown: (CancellationToken, mpsc::Sender<()>),
) -> JoinHandle<Result<ExitStatus, Errors>> {
    config.boot_peers = genesis
        .boot_peers(Some(topos_p2p::constants::TCE_BOOTNODE_PORT))
        .into_iter()
        .chain(config.parse_boot_peers())
        .collect::<Vec<_>>();
    config.auth_key = keys.network.map(AuthKey::PrivateKey);
    config.signing_key = keys.validator.map(AuthKey::PrivateKey);
    config.p2p.is_bootnode = if let Some(AuthKey::PrivateKey(ref k)) = config.auth_key {
        let peer_id = topos_p2p::utils::keypair_from_protobuf_encoding(&k[..])
            .public()
            .to_peer_id();

        config.boot_peers.iter().any(|(p, _)| p == &peer_id)
    } else {
        false
    };

    config.validators = genesis.validators().expect("Cannot parse validators");
    config.tce_params = ReliableBroadcastParams::new(config.validators.len());

    if let Some(socket) = config.libp2p_api_addr {
        warn!(
            "`libp2p_api_addr` is deprecated in favor of `listen_addresses` and \
             `public_addresses` and will be removed in the next version. In order to keep your \
             node running, `libp2p_api_addr` will be used."
        );

        let addr: Multiaddr = format!("/ip4/{}/tcp/{}", socket.ip(), socket.port())
            .parse()
            .expect("Unable to generate Multiaddr from `libp2p_api_addr`");

        config.p2p.listen_addresses = vec![addr.clone()];
        config.p2p.public_addresses = vec![addr];
    }

    config.version = env!("TOPOS_VERSION");
    config.storage = StorageConfiguration::RocksDB(Some(config.db_path.clone()));

    debug!("TCE args: {config:?}");
    spawn(async move {
        topos_tce::run(&config, shutdown).await.map_err(|e| {
            error!("TCE process terminated: {e:?}");
            Errors::TceFailure
        })
    })
}

pub fn spawn_edge_process(
    edge_path: PathBuf,
    data_dir: PathBuf,
    genesis_path: PathBuf,
    edge_args: HashMap<String, String>,
) -> JoinHandle<Result<ExitStatus, Errors>> {
    spawn(async move {
        CommandConfig::new(edge_path)
            .server(&data_dir, &genesis_path, edge_args)
            .spawn()
            .await
            .map_err(Errors::EdgeTerminated)
    })
}
