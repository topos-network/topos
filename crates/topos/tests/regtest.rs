mod utils;

use std::process::Command;

use assert_cmd::prelude::*;

#[test]
fn regtest_spam_help_display() -> Result<(), Box<dyn std::error::Error>> {
    let mut cmd = Command::cargo_bin("topos")?;
    cmd.arg("regtest").arg("spam").arg("-h");

    let output = cmd.assert().success();

    let result: &str = std::str::from_utf8(&output.get_output().stdout)?;

    insta::assert_snapshot!(utils::sanitize_config_folder_path(result));

    Ok(())
}

#[test]
fn regtest_spam_invalid_hosts() -> Result<(), Box<dyn std::error::Error>> {
    let mut cmd = Command::cargo_bin("topos")?;
    cmd.arg("regtest")
        .arg("spam")
        .arg("--benchmark")
        .arg("--target-hosts")
        .arg("asd")
        .arg("--number")
        .arg("1");

    let output = cmd.assert().failure();

    let result: &str = std::str::from_utf8(&output.get_output().stdout)?;

    assert!(result.contains(
        "Invalid target-hosts pattern. Has to be in the format of http://validator-1:9090"
    ));

    Ok(())
}

#[test]
fn regtest_spam_invalid_number() -> Result<(), Box<dyn std::error::Error>> {
    let mut cmd = Command::cargo_bin("topos")?;
    cmd.arg("regtest")
        .arg("spam")
        .arg("--benchmark")
        .arg("--target-hosts")
        .arg(" http://validator-{N}:9090")
        .arg("--number")
        .arg("dasd");

    cmd.assert().failure();

    Ok(())
}
