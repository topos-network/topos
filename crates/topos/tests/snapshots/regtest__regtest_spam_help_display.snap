---
source: crates/topos/tests/regtest.rs
expression: "utils::sanitize_config_folder_path(result)"
---
Run a test topos certificate spammer to send test certificates to the network, generating randomly among the `nb_subnets` subnets the batch of `cert_per_batch` certificates at every `batch-interval`

Usage: topos regtest spam [OPTIONS]

Options:
      --target-nodes <TARGET_NODES>
          The target node api endpoint. Multiple nodes could be specified as comma separated list e.g. `--target-nodes=http://[::1]:1340,http://[::1]:1341` [env: TOPOS_NETWORK_SPAMMER_TARGET_NODES=]
  -v, --verbose...
          Defines the verbosity level
      --no-color
          Disable color in logs [env: TOPOS_LOG_NOCOLOR=]
      --target-nodes-path <TARGET_NODES_PATH>
          Path to json file with list of target nodes as alternative to `--target-nodes` [env: TOPOS_NETWORK_SPAMMER_TARGET_NODES_PATH=]
      --home <HOME>
          Home directory for the configuration [env: TOPOS_HOME=] [default: /home/runner/.config/topos]
      --local-key-seed <LOCAL_KEY_SEED>
          Seed for generation of local private signing keys and corresponding subnet ids [env: TOPOS_NETWORK_SPAMMER_LOCAL_KEY_SEED=] [default: 1]
      --cert-per-batch <CERT_PER_BATCH>
          Certificates generated in one batch. Batch is generated every `batch-interval` milliseconds [env: TOPOS_NETWORK_SPAMMER_CERT_PER_BATCH=] [default: 1]
      --nb-subnets <NB_SUBNETS>
          Number of subnets to use for certificate generation. For every certificate subnet id will be picked randomly [env: TOPOS_NETWORK_SPAMMER_NUMBER_OF_SUBNETS=] [default: 1]
      --nb-batches <NB_BATCHES>
          Number of batches to generate before finishing execution. If not specified, batches will be generated indefinitely [env: TOPOS_NETWORK_SPAMMER_NUMBER_OF_BATCHES=]
      --batch-interval <BATCH_INTERVAL>
          Time interval in milliseconds between generated batches of certificates [env: TOPOS_NETWORK_SPAMMER_BATCH_INTERVAL=] [default: 2000]
      --target-subnets <TARGET_SUBNETS>
          List of generated certificate target subnets. No target subnets by default. For example `--target-subnets=0x3bc19e36ff1673910575b6727a974a9abd80c9a875d41ab3e2648dbfb9e4b518,0xa00d60b2b408c2a14c5d70cdd2c205db8985ef737a7e55ad20ea32cc9e7c417c` [env: TOPOS_NETWORK_SPAMMER_TARGET_SUBNETS=]
      --otlp-agent <OTLP_AGENT>
          Socket of the opentelemetry agent endpoint. If not provided open telemetry will not be used [env: TOPOS_OTLP_AGENT=]
      --otlp-service-name <OTLP_SERVICE_NAME>
          Otlp service name. If not provided open telemetry will not be used [env: TOPOS_OTLP_SERVICE_NAME=]
  -h, --help
          Print help

